# Deriving New Variables

## Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

## Importing Data

In [2]:
# Assigning the path

path = r'C:\Users\guest23455\Desktop\CareerFoundry\Data Immersion\Achievement 4_Python Fundamentals for Data Analysts\05-2023 Instacart Basket Analysis'

In [3]:
# Import ords_prods_merged

df_ords_prods_merged =  pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data','orders_products_merged.pkl'))

In [4]:
# Create a subset with the first million rows from df_ords_prods_merged (orders_products_merged.pkl)

df = df_ords_prods_merged[:1000000]

In [5]:
df.shape

(1000000, 14)

## Using If-Statements 

### With User-Defined Functions

In [8]:
# Define the price label function

def price_label(row): 
    
    if row['prices'] <= 5:                                   
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):     
        return 'Mid-range product'
    elif row['prices'] > 15:                                   
        return 'High-range product'
    else : return 'Not enough data'
    

# If the item’s price is lower than or equal to $5, it will be labeled a “low-range product”
# If the item’s price is above $5 but lower than or equal to $15, it will be labeled a “mid-range product”
# If the item’s price is above $15, it will be labeled a “high-range product”

In [9]:
# Apply the price label functionon the dataframe

df['price range'] = df.apply(price_label, axis = 1)

C:\Users\guest23455\AppData\Local\Temp\ipykernel_17176\3097047293.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price range'] = df.apply(price_label, axis = 1)


In [10]:
# Check value counts

df['price range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price range, dtype: int64

In [11]:
# The most expensive product within the subset

df['prices'].max()

14.8

### With loc() Function

In [14]:
# If the item’s price is above $15, it will be labeled a “high-range product”

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [13]:
# If the item’s price is above $5 but lower than or equal to $15, it will be labeled a “mid-range product”

df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [15]:
# If the item’s price is lower than or equal to $5, it will be labeled a “low-range product”

df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [16]:
# Check value counts

df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

Repeat the process on the entire dataframe and not just the subset

In [17]:
# If the item’s price is above $15, it will be labeled a “high-range product”

df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, ' price_range_loc'] = 'High-range product'

In [18]:
# If the item’s price is above $5 but lower than or equal to $15, it will be labeled a “mid-range product”

df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [19]:
# If the item’s price is lower than or equal to $5, it will be labeled a “low-range product”

df_ords_prods_merged.loc[df_ords_prods_merged['prices'] < 5, 'price_range_loc'] = "Low-range product"

In [20]:
# Check value counts

df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product    21860860
Low-range product     9900417
NaN                    643582
Name: price_range_loc, dtype: int64

### With For-Loops

In [21]:
# Check value counts 

# 0 - Saturday , 1 - Sunday, 2 - Monday, 3 - Tuesday, 4 - Wednesday, 5 - Thursday, 6 - Friday 

df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

# Frequency of value '0' is the highest, which means Saturday is the busiest day of the week 
# Frequency of value '4' is the lowest, which means Wednesday is the slowest (least busy) day of the week

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

**Create a new column called 'Busiest day', which will contain three different values : 'Busiest day', 'Least busy', and 'Regularly busy'**


In [24]:
# Create an empty list called result

# Set loop

result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [25]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [26]:
# Combine the list with the newly created column

df_ords_prods_merged['busiest_day'] = result

In [27]:
# Check value counts

df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

# Task 2 : New column for busiest days

In [28]:
# Create an empty list called result1

# Set loop

result1 = []

for value in df_ords_prods_merged["orders_day_of_week"]:
    if value == 0 or value == 1 :
        result1.append("Busiest day")
    elif value == 4 or value == 3:
        result1.append("Least busy")
    else:
        result1.append("Regularly busy")

In [29]:
# Combine the list with the newly created column

df_ords_prods_merged['busiest_days'] = result1

In [32]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    0.0         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  price_range_loc  \
0         Soda        77              7     9.0   both              NaN   
1         Soda        77              7     9.0   both              NaN   
2         Soda        77              7     9.0   both              NaN   
3         Soda        77              7     9.0   both              NaN   
4         Soda        77              7     9.0   both              NaN   

     price_range_loc     busiest_day    busiest_days  
0  Mid-range product  Regularly busy  Regularly busy  
1  Mid-range product  Regularly busy      Least busy  
2  Mid-range product  Regularly busy      Least busy  
3  Mid-range product      Least busy      Least busy  
4  Mid-range product      Least busy      Least busy

# Task 3 : Check for accuracy

In [33]:
# Check value counts

df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest day       11864412
Least busy         7624336
Name: busiest_days, dtype: int64

Busiest day is 0 (Saturday) + 1 (Sunday) => 6204182 + 5660230 = 11864412 </br>
Regular busy is 2 (Monday) + 5 (Thursday) + 6 (Friday) =>  4496490 + 4213830 + 4205791 = 12916111 </br>
Least busy is 3 (Tuesday) + 4 (Wednesday) => 3840534 + 3783802 => 7624336 </br>

# Task 4 : Busiest period of day

In [34]:
df_ords_prods_merged['order_hour_of_day'].value_counts()

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

Most orders = 10, 11, 14, 15, 13, 12, 16, 9, 17 (orders above 2000000) </br>
Average orders = 8, 18, 19, 20, 7, 21, 22 (orders above 500000) </br>
Fewest orders = 23, 6, 0, 1, 5, 2, 4, 3 

In [35]:
# Create an empty list called hod (hour of day)

# Set loop

hod = []

for value in df_ords_prods_merged["order_hour_of_day"]:
    if value in [10, 11, 14, 15, 13, 12, 16, 9, 17] :
        hod.append("Most orders")
    elif value in [ 8, 18, 19, 20, 7, 21, 22]:
        hod.append("Average orders")
    else:
        hod.append("Fewest orders")

In [36]:
# Combine the list with the newly created column

df_ords_prods_merged['busiest_period_of_day'] = hod

In [38]:
# Verify if the results are correct

df_ords_prods_merged[['order_hour_of_day', 'busiest_period_of_day']].tail()

order_hour_of_day busiest_period_of_day
32404854                 15           Most orders
32404855                 11           Most orders
32404856                  7        Average orders
32404857                  5         Fewest orders
32404858                 14           Most orders

# Task 5 : Print the frequency of busiest_period_of_day

In [39]:
df_ords_prods_merged['busiest_period_of_day'].value_counts()

Most orders       23205725
Average orders     7909997
Fewest orders      1289137
Name: busiest_period_of_day, dtype: int64

# Task 6 : Export the dataframe as a pickle file

In [40]:
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated.pkl'))